evm/p2p/les.py

```diff
@@ -6,14 +18,159 @@
+            # The sedes.binary we use in .structure above will give us a bytes value here, but
+            # using bytes as dictionary keys makes it impossible to use the dict() constructor
+            # with keyword arguments, so we convert them to strings here.
+            key = key.decode('ascii')
+            if key not in self.items_sedes:
+                continue
+            item_sedes = self.items_sedes[key]
+            if item_sedes is not None:
+                decoded[key] = item_sedes.deserialize(value)
+            else:
+                decoded[key] = value
+        return decoded
+
+    def encode_payload(self, data):
+        response = []
+        for key, value in data.items():
+            response.append((key, self.items_sedes[key].serialize(value)))
+        return super(Status, self).encode_payload(response)
+
+    def handle(self, proto, data):
+        return self.decode(data)
+
+
+class Announce(Command):
+    _cmd_id = 1
+    structure = [
+        ('headHash', sedes.binary),
+        ('headNumber', sedes.big_endian_int),
+        ('headTd', sedes.big_endian_int),
+        ('reorgDepth', sedes.big_endian_int),
+        ('params', sedes.CountableList(sedes.List([sedes.binary, sedes.raw]))),
+    ]
+    # TODO: The params CountableList above may contain any of the values from the Status msg.
+    # Need to extend this command to process that too.
+
+    def handle(self, proto, data):
+        decoded = self.decode(data)
+        # FIXME: Should ask only for block headers we don't have yet.
+        proto.send_get_block_headers(decoded['headNumber'])
+        return decoded
+
+
+class GetBlockHeadersQuery(rlp.Serializable):
+    fields = [
+        # TODO: It should be possible to specify the block either by its number or hash, but
+        # for now only the number is supported.
+        ('block', sedes.big_endian_int),
+        ('maxHeaders', sedes.big_endian_int),
+        ('skip', sedes.big_endian_int),
+        ('reverse', sedes.big_endian_int),
+    ]
+
+
+class GetBlockHeaders(Command):
+    _cmd_id = 2
+    structure = [
+        ('request_id', sedes.big_endian_int),
+        ('query', GetBlockHeadersQuery),
+    ]
+
+
+class BlockHeaders(Command):
+    _cmd_id = 3
+    structure = [
+        ('request_id', sedes.big_endian_int),
+        ('buffer_value', sedes.big_endian_int),
+        ('headers', sedes.CountableList(BlockHeader)),
+    ]
 
     def handle(self, proto, data):
-        proto.logger.debug("Got LES/Status msg")
+        decoded = self.decode(data)
+        last_header_seen = decoded['headers'][0].block_number
+        proto.logger.debug("Last block header received: {}".format(last_header_seen))
+        return decoded
 
 
 class LESProtocol(Protocol):
     name = b'les'
     version = 1
-    _commands = [Status]
-    # FIXME: Need to find out the correct value for this
-    cmd_length = 21
+    _commands = [Status, Announce, BlockHeaders]
+    _req_id = 0
+    max_headers = 10
```

Comment:
Where does this number come from?

Reply:
It's arbitrary, just the # of block headers we ask for when we send a GetBlockHeaders request. I'll add a FIXME comment